### DISCUSSION TITLE
Validating sources as containing a gain map?

### DISCUSSION THREAD

**Discussion by gregbenz** [google/libultrahdr#137] [Q&A] [answered]
Does ultrahdr_app have a feature which might be used simply to validate that an image contains a gain map? For example, this could be useful to triage uploads to a website.

If this isn't supported, please consider this as a feature request for the future to support efficient queries for gain maps natively supported by the project.

**Comment by mcDandy** [comment]
The app does not have a validation only mode. The library itself however
has it as a functions (but my understanding to the library Is very limited)
but you cannot use library itself.

Dne po 13. 5. 2024 19:35 uživatel Greg B ***@***.***> napsal:

**Reply by ram-mohan** [reply]
The sample application does not have an option to determine if the image is a valid uhdr. But the library interface exposes functions that can determine the same.

  uhdr_codec_private_t* obj = uhdr_create_decoder();  // create decoder
  uhdr_compressed_image_t uhdr_image{};
  uhdr_image.data = img->data;
  uhdr_image.data_sz = img->length;
  uhdr_image.capacity = img->length;
  uhdr_image.cg = UHDR_CG_UNSPECIFIED;
  uhdr_image.ct = UHDR_CT_UNSPECIFIED;
  uhdr_image.range = UHDR_CR_UNSPECIFIED;
  RET_IF_ERR(uhdr_dec_set_image(obj, &uhdr_image))  // register image
  RET_IF_ERR(uhdr_probe(obj))  // probe image
  uhdr_release_decoder(obj);  // release decoder

**Reply by ram-mohan** [reply]
If probe call is successful, there are variety of function calls that can be made after probe to get useful data, like 

uhdr_dec_get_image_width(obj) returns primary image width
uhdr_dec_get_image_height(obj) returns primary image height
uhdr_dec_get_gainmap_width(obj) returns gainmap image width
uhdr_dec_get_gainmap_height(obj) returns gainmap image height
uhdr_dec_get_gain_map_metadata(obj) returns gainmap metadata

**Reply by gregbenz** [reply]
@ram-mohan Thank you. What would you recommend as the most performant way to use the above functions to determine that the target had a valid gain map? Ultimately, I would think metadata checks of the parent/auxiliary image would be suitable (ie, check headers/metadata and assume any images present are not corrupt)

**Reply by ram-mohan** [reply, marked as answer]
uhdr_dec_probe(), checks if there is a valid primary image, gainmap image and valid hdr metadata. If it returns success, then we can assume its a valid uhdr image. Primary image / gainmap image being corrupt (headers are ok, but entropy coded residue data is corrupted) is realized only during actual decode. 

uhdr_dec_probe() return status should be a good starting filtering check.

**Reply by gregbenz** [reply]
Thank you, @ram-mohan !
